Define character traits for `std::uint8_t`

ChrisThrasher opened the pull request:
## Description

Closes #2670 

Character traits are only standardized for character types of which std::uint8_t is not. All major C++ implementations happen to define this specialization but because it is not standard C++ they are allowed to remove it as LLVM has done by deprecating this specialization in LLVM 18. It is slated for removal in LLVM 19. To avoid compilation errors and to get ahead of any deprecation warnings when LLVM 18 ships we need to define our own specialization of std::uint8_t character traits.

SFML 4 will have access to C++20's std::u8string which should let us remove this code.

Quuxplusone commented:
The commit summary still says "`std::char_traits`" when it means something like "a user-defined traits class"; but other than that, LGTM!

ChrisThrasher commented:
Thanks I’ll clean up that comment

ChrisThrasher commented on src/SFML/System/String.cpp:
```diff
@@ -37,6 +37,43 @@
 
 namespace sf
 {
+////////////////////////////////////////////////////////////
```
SFML tries hard to keep headers minimal so I implemented these functions in the source file. I’m open to reevaluating that if there are convincing benchmarks to the contrary. 

Because this header is used in many places there are compile time benefits to not inlining the functions.

philnik777 replied:
It's not a benchmark, but maybe it's convincing enough: https://godbolt.org/z/4Mb3Mxhcs. Also note that these functions are really small and thus very cheap to instantiate.

ChrisThrasher commented:
@JohelEGP Can you test if this fixes your issue?

philnik777 commented on include/SFML/System/String.hpp:
```diff
@@ -39,6 +39,39 @@
+    ///
+    ////////////////////////////////////////////////////////////
+    using char_type = std::uint8_t;
+
```
You may not need them right now, but it's also not great if it breaks down the road because you didn't add all the required functionality, especially since it's trivial to add right now. A string implementation is free to use all of it.

ChrisThrasher replied:
If you think it's trivial then would you mind adding the remaining missing functions?

JohelEGP reviewed (approved):
Thank you!
Now building SFML doesn't warn.
Although I'm using <https://github.com/SFML/imgui-sfml/>,
so I can't tell whether my project will build until that's updated.

ChrisThrasher commented:
I implemented all the remaining `std::char_traits` aliases and functions. I'm not sure if they're actually necessarily but I implemented them out of an abundance of caution.

kimci86 commented on test/System/String.test.cpp:
```diff
@@ -58,6 +58,134 @@ struct StringMaker<std::basic_string<std::uint8_t>>
+
+    SECTION("not_eof()")
+    {
+        CHECK(!sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{0}));
+        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{1}));
+        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{10}));
+        CHECK(!sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::eof()));
```
This function is supposed to return the argument when it is not eof, let's test that:
```suggestion
        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{0}) == sf::U8StringCharTraits::int_type{0});
        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{1}) == sf::U8StringCharTraits::int_type{1});
        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::int_type{10}) == sf::U8StringCharTraits::int_type{10});
        CHECK(sf::U8StringCharTraits::not_eof(sf::U8StringCharTraits::eof()) != sf::U8StringCharTraits::eof());
```

Besides, not sure if being explicit about `sf::U8StringCharTraits::int_type` is useful and it is quite verbose.

kimci86 commented on test/System/String.test.cpp:
```diff
@@ -58,6 +58,140 @@ struct StringMaker<std::basic_string<std::uint8_t>>
+    SECTION("assign(char_type*, std::size_t, char_type)")
+    {
+        sf::U8StringCharTraits::char_type s[] = {'a', 'b', 'c', '\0'};
+        CHECK(sf::U8StringCharTraits::assign(s, 2, 'd') == &s[0]);
```
```suggestion
        CHECK(sf::U8StringCharTraits::assign(s, 2, 'd') == &s[0]);
        CHECK(s[0] == 'd');
        CHECK(s[1] == 'd');
        CHECK(s[2] == 'c');
```

kimci86 commented on src/SFML/System/String.cpp:
```diff
@@ -37,6 +37,108 @@
+////////////////////////////////////////////////////////////
+U8StringCharTraits::int_type U8StringCharTraits::to_int_type(char_type c) noexcept
+{
+    return std::char_traits<char>::to_int_type(static_cast<std::char_traits<char>::char_type>(c));
```
```suggestion
    return std::char_traits<char>::to_int_type(static_cast<char>(c));
```

kimci86 reviewed (approved):
LGTM besides nitpicking.

ChrisThrasher commented:
https://github.com/llvm/llvm-project/commit/c3668779c13596e223c26fbd49670d18cd638c40

LLVM has removed this base template. We merged this just in time!
